extensions/metadata.md

```diff
@@ -0,0 +1,739 @@
+
+`Value` is a UTF-8 encoded value.
+
+## METADATA client command
+
+    METADATA <Target> <Subcommand> [<Param 1> ... [<Param n>]]
+
+`Target` is a valid nickname or channel name. Clients SHOULD use the asterisk symbol (`*`) when targeting their own nickname.
+
+`Subcommand` is one of the subcommands listed below. The allowed params are described in each subcommand description.
+
+### METADATA GET
+
+    METADATA <Target> GET key1 key2 ...
+
+This subcommand lets clients lookup keys on the given target.
+
+Multiple keys may be given.
+
+The response will be a `metadata` batch containing:
+
+* `RPL_KEYVALUE`
+* `RPL_KEYNOTSET`
+* `FAIL METADATA KEY_INVALID`
+
+for every key in order.
+
+Servers MAY replace metadata which is considered not visible for the requesting user, with `RPL_KEYNOTSET` or with `FAIL METADATA KEY_NO_PERMISSION`.
+
+*Failures*: `INVALID_KEY`, `KEY_NO_PERMISSION`
+
+### METADATA LIST
+
+    METADATA <Target> LIST
+
+This subcommand lists all of the target's currently-set metadata keys along with their values.
+
+If the target is valid, the response is a `metadata` batch containing zero or more `RPL_KEYVALUE` events.
+If the target is not valid, ONLY a `FAIL METADATA INVALID_TARGET` reply is sent.
+
+Servers MAY omit metadata which is considered not visible for the requesting user, or replace it with `FAIL METADATA KEY_NO_PERMISSION` within the batch.
+
+*Failures*: `FAIL METADATA KEY_NO_PERMISSION`.
+
+### METADATA SET
+
+    METADATA <Target> SET <Key> [:Value]
+
+This subcommand sets the key on the target to the given value. If no value is given, the key is removed.
+
+If the key is invalid, the server responsds with `FAIL METADATA KEY_INVALID` and fails the request.
+
+If the key is valid, but not set, and the client tries to remove the key, the server responds with `FAIL METADATA KEY_NOT_SET` and fails the request.
+
+If the user cannot set keys on the given target, the server responds with `FAIL METADATA KEY_NO_PERMISSION` and fails the request.
+
+Servers MAY respond to certain keys considered not settable by the requesting user, or otherwise disallowed by the server, with `FAIL METADATA KEY_NO_PERMISSION` and fail the request.
+
+Servers MAY respond with `FAIL METADATA RATE_LIMITED` and fail the request. When a client receives `FAIL METADATA RATE_LIMITED`, it SHOULD retry the `METADATA SET` request at a later time. If the `FAIL METADATA RATE_LIMITED` event contains the `<RetryAfter>` parameter, the parameter value MUST be a positive integer indicating the minimum number of seconds the client should wait before retrying the request.
+
+If the request is successful, the server carries out the requested change and responds with one `RPL_KEYVALUE` event, representing the new value (or lack of one).
```

Comment:
It might be useful to note that the server tweak the value and that clients should check the response before storing it in any local cache.